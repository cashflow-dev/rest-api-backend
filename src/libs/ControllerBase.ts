/* eslint-disable no-param-reassign */
import { ParameterizedContext } from 'koa';
import { serviceLaunch } from '../services';
import { SuccessBody } from '../interfaces/SuccessBody';
import { handleHttpErrors } from './utils';
import HTTP_STATUS_CODE from '../enums/HTTP_STATUS_CODE';

const ControllerBase = (service: any) => {
  const controller: any = {};

  controller.create = async (context: ParameterizedContext): Promise<void> => {
    try {
      await serviceLaunch(service, 'create', context);
      context.status = HTTP_STATUS_CODE.CREATED;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  controller.find = async (context: ParameterizedContext): Promise<void> => {
    try {
      const { users, next } = await serviceLaunch(service, 'find', context);
      context.staus = HTTP_STATUS_CODE.OK;
      const body: SuccessBody = {
        message: 'ok',
        data: users,
        next,
        statusCode: HTTP_STATUS_CODE.OK,
      };
      context.body = body;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  controller.findById = async (context: ParameterizedContext): Promise<void> => {
    try {
      const user = await serviceLaunch(service, 'findById', context);
      context.staus = HTTP_STATUS_CODE.OK;
      const body: SuccessBody = {
        message: 'ok',
        data: user,
        statusCode: HTTP_STATUS_CODE.OK,
      };
      context.body = body;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  controller.update = async (context: ParameterizedContext): Promise<void> => {
    try {
      await serviceLaunch(service, 'update', context);
      context.status = HTTP_STATUS_CODE.NO_CONTENT;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  controller.deleteById = async (context: ParameterizedContext): Promise<void> => {
    try {
      await serviceLaunch(service, 'delete', context);
      context.status = HTTP_STATUS_CODE.NO_CONTENT;
    } catch (e) {
      handleHttpErrors(e, context);
    }
  };

  return controller;
};

export default ControllerBase;
