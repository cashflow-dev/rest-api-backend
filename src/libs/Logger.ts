import moment from 'moment';
import bunyan from 'bunyan';
import chalk from 'chalk';
import 'dotenv/config';

const log = bunyan.createLogger({
  name: `${process.env.APP_NAME}`,
  src: true,
  streams: [
    {
      type: 'file',
      path: `${__dirname}/..${process.env.LOG_PATH}`,
      level: 'info',
    },
    {
      type: 'file',
      path: `${__dirname}/..${process.env.LOG_PATH}`,
      level: 'warn',
    },
    {
      type: 'file',
      path: `${__dirname}/..${process.env.LOG_PATH}`,
      level: 'error',
    },
    {
      type: 'file',
      path: `${__dirname}/..${process.env.LOG_PATH}`,
      level: 'fatal',
    },
  ],
});

class Logger {
  public static info(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    console.log(chalk.inverse.bold(logMessage));
    log.info(logMessage);
  }

  public static warn(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    console.log(chalk.bgYellow.black.bold(logMessage));
    log.warn(logMessage);
  }

  public static error(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    console.log(chalk.bgRed.bold.white(logMessage));
    log.error(logMessage);
  }

  public static vebose(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    log.info(logMessage);
  }

  public static debug(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    log.debug(logMessage);
  }

  public static fatal(msg: string) {
    const time = moment().format('YYYY-MM-DD h:mm:ss a');
    const logMessage = `${time}: ${msg}`;
    console.log(chalk.bgRed.bold.white(logMessage));
    log.fatal(logMessage);
  }
}

export default Logger;
