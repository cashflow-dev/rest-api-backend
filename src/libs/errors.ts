export class NotFoundError extends Error {
  public constructor() {
    super('NotFoundError');
    Error.captureStackTrace(this, NotFoundError);
  }
}

export class UnauthorizedError extends Error {
  public constructor() {
    super('Unauthorized');
    Error.captureStackTrace(this, UnauthorizedError);
  }
}

export class ValidationError extends Error {
  public validationMessages: string[];

  public constructor(validationMessages: string[]) {
    super('ValidationError');
    this.validationMessages = validationMessages;
    Error.captureStackTrace(this, ValidationError);
  }
}
