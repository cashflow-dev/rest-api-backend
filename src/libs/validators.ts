import Joi from '@hapi/joi';
import joiObjectId from 'joi-objectid';
import _ from 'lodash';
import { InputData } from '../interfaces/InputData';
import ValidationError from '../helpers/ValidationError';

const ObjectId = joiObjectId(Joi);

export const composeValidators = (...functions: ((inputData: InputData) => void)[]) => {
  return function(inputData: InputData) {
    functions.forEach(func => {
      func(inputData);
    });
  };
};

export const validateId = (inputData: InputData) => {
  const schema = {
    params: {
      id: ObjectId()
        .required()
        .error(() => 'Not a valid id'),
    },
  };

  const paramsSchema = Joi.object().keys(schema);

  const result = Joi.validate(inputData, paramsSchema, { stripUnknown: { arrays: true, objects: true } });
  if (result.error) {
    const validationMessages = result.error.details.map(error => error.message);
    throw new ValidationError(validationMessages);
  }
};

export const validateSearch = (inputData: InputData) => {
  const schema = {
    query: {
      search: Joi.string(),
    },
  };

  const querySchema = Joi.object().keys(schema);

  const result = Joi.validate(inputData, querySchema, { stripUnknown: { arrays: true, objects: true } });
  if (result.error) {
    const validationMessages = result.error.details.map(error => error.message);
    throw new ValidationError(validationMessages);
  }
};
export const validateQuery = (inputData: InputData) => {
  const schema = {
    query: {
      next: ObjectId().error(() => 'Not a valid cursor'),
      fields: Joi.string(),
      limit: Joi.number().max(50),
    },
  };

  const querySchema = Joi.object().keys(schema);

  const result = Joi.validate(inputData, querySchema, { stripUnknown: { arrays: true, objects: true } });
  if (result.error) {
    const validationMessages = result.error.details.map(error => error.message);
    throw new ValidationError(validationMessages);
  }
};
