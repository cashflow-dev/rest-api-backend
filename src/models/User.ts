import { Schema, model, Document } from 'mongoose';
import { bcrypt } from 'bcrypt';
import { timestamps } from 'mongoose-timestamp';

const saltRounds = 10;

export interface User extends Document {
  email: string;
  password: string;
  currency: Currency;
  settings: Settings;
}

interface Settings extends Document {
  currency: string;
}

interface Currency extends Document {
  origin: string;
}

const userSchema = new Schema({
  email: { type: String, unique: true },
  password: String,
  currency: {
    origin: String,
  },
  settings: {
    currency: String,
  },
});

userSchema.pre('save', function(next) {
  const user = this;

  if (!user.isModified('password')) return next();

  bcrypt.genSalt(saltRounds, function(err, salt) {
    if (err) return next(err);

    bcrypt.hash(user.password, salt, function(error, hash) {
      if (error) return next(error);

      user.password = hash;
      next();
      return null;
    });
    return null;
  });
  return null;
});

userSchema.methods.comparePassword = function(candidatePassword, cb) {
  bcrypt.compare(candidatePassword, this.password, function(err, isMatch) {
    if (err) return cb(err);
    cb(null, isMatch);
    return null;
  });
};

userSchema.plugin(timestamps);

export default model<User>('User', userSchema);
