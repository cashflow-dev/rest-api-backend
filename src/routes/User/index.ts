import Router from 'koa-router';
import UserController from '../../controllers/user';

const router = new Router({ prefix: '/user' });
const controller = UserController();

router.get('/:id', controller.findById);
router.post('/', controller.create);
router.patch('/:id', controller.update);
router.get('/', controller.find);
router.delete('/:id', controller.deleteById);

export default router;
