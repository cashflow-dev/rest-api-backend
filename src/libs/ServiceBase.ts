import _ from 'lodash';
import { InputData } from '../interfaces/InputData';
import { Service } from '../interfaces/Service';
import { composeValidators, validateId, validateQuery } from './validators';
import { NotFoundError } from './errors';
import { getRequestedFields, handleMongoErrors } from './utils';

export class ServiceBase implements Service {
  public input: InputData;

  public DEFAULT_LIMIT: number;

  public Model: any;

  public requestedFields: any;

  public queryFilter: any;

  public validators: any;

  public constructor(input: InputData) {
    this.DEFAULT_LIMIT = 50;
    this.input = input;
    this.validators = {
      find: composeValidators(validateQuery),
      findById: composeValidators(validateId),
      delete: composeValidators(validateId),
    };

    /* eslint-disable no-underscore-dangle */
    this.requestedFields = this.input.fields ? getRequestedFields(this.input.fields) : { _id: 1 };
    this.queryFilter = this.input.next
      ? {
          _id: { $lt: this.input.next },
        }
      : {};
  }

  public async findById(): Promise<any> {
    let result;
    try {
      result = await this.Model.findById(this.input.params.id).select(this.requestedFields);
    } catch (e) {
      handleMongoErrors(e);
    }
    if (result === null) {
      throw new NotFoundError();
    }
    return result;
  }

  public async create(): Promise<any> {
    let result;
    try {
      const model = new this.Model(this.input.body);
      result = await model.save();
    } catch (e) {
      handleMongoErrors(e);
    }
    return result;
  }

  public async find(): Promise<any> {
    let results: any;
    try {
      results = await this.Model.find(this.queryFilter)
        .select(this.requestedFields)
        .sort({ _id: -1 })
        .limit(Number(this.input.limit) || this.DEFAULT_LIMIT);
    } catch (e) {
      handleMongoErrors(e);
    }
    return { users: results, next: _.get(results[results.length - 1], '_id', null) };
  }

  public async update(): Promise<any> {
    let result = null;
    try {
      result = await this.Model.findByIdAndUpdate(this.input.params.id, this.input.body);
    } catch (e) {
      handleMongoErrors(e);
    }
    if (result === null) {
      throw new NotFoundError();
    }
    return result;
  }

  public async delete(): Promise<any> {
    let result = null;
    try {
      result = await this.Model.findByIdAndDelete(this.input.params.id);
    } catch (e) {
      handleMongoErrors(e);
    }
    if (result === null) {
      throw new NotFoundError();
    }
    return result;
  }
}

export default ServiceBase;
