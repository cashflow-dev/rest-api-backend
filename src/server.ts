import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import logger from 'koa-logger';
import cors from '@koa/cors';
import debug from 'debug';
import fs from 'fs';
import router from './routes/index';
import { connect } from './models';
import Logger from './libs/Logger';
import 'dotenv/config';

const LOG_PATH = `${__dirname}/${process.env.LOG_PATH}`;

if (!fs.existsSync(LOG_PATH)) {
  fs.closeSync(fs.openSync(LOG_PATH, 'w'));
}

const DATABASE = `${process.env.MONGODB_URI}`;
console.log({ DATABASE });

const PORT = 3003;

connect(
  DATABASE,
  Logger.fatal,
);
const app = new Koa();
app.use(cors());
app.use(bodyParser());
app.use(logger());
app.use(router());

app.listen(PORT, (): void => {
  debug(`API Server listening on port: ${PORT}`);
});
