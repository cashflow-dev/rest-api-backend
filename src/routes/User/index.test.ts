import request from 'supertest';
import _ from 'lodash';
import { app } from '../../server';
import { stripIdAndNext } from '../../libs/utils';
import User from '../../models/User';

let server: any;

describe('/user', () => {
  let userId: any;

  beforeAll(async () => {
    server = app.listen();
    const promises = [];
    for (let i = 0; i < 5; i += 1) {
      const user = new User({ email: `test${i}@test.com`, password: `test${i}` });
      promises.push(user.save());
    }
    const res = await Promise.all(promises);
    /* eslint-disable no-underscore-dangle */
    userId = res[0]._id;
    /* eslint-enable no-underscore-dangle */
  });

  afterAll(async () => {
    await server.close();
  });
  describe('GET /user', () => {
    it('should get users', async () => {
      const response = await request(server).get('/user?fields=email,password');
      expect(stripIdAndNext(response.text)).toMatchSnapshot();
      expect(response.status).toBe(200);
      expect(JSON.parse(response.text).data.length).toBeGreaterThan(0);
    });
  });
  describe('POST /user', () => {
    describe('with all fields', () => {
      const user = { email: 'test@test.com', password: 'test' };
      let response: any;
      let dbUser: any;

      beforeAll(async () => {
        response = await request(server)
          .post('/user')
          .send(user);
        dbUser = await User.find(user);
      });
      it('should match snapshot', async () => {
        expect(response.text).toMatchSnapshot();
        expect(response.status).toBe(201);
      });
      it('should create a new user in the database', () => {
        expect(dbUser.length).toBe(1);
      });
    });
    describe('with fields missing', () => {
      describe('with email missing', () => {
        it('should match snapshot', async () => {
          const user = { password: 'test' };
          const response = await request(server)
            .post('/user')
            .send(user);
          expect(response.text).toMatchSnapshot();
          expect(response.status).toBe(400);
        });
      });
      describe('with password missing', () => {
        it('should match snapshot', async () => {
          const user = { email: 'test@test.com' };
          const response = await request(server)
            .post('/user')
            .send(user);
          expect(response.text).toMatchSnapshot();
          expect(response.status).toBe(400);
        });
      });
    });
  });
  describe('PATCH /user/:id -> Update password', () => {
    let updateData: any;
    let initUser: any;
    let response: any;
    let updatedUser: any;
    beforeAll(async () => {
      const query = { _id: userId };
      updateData = { password: 'newTestPassword' };
      initUser = await User.find(query);
      response = await request(server)
        .patch(`/user/${userId}`)
        .send(updateData);
      updatedUser = await User.find(query);
    });
    it('should currently have another password', () => {
      expect(initUser[0].password === updateData.password).toBe(false);
    });
    it('should respond with 204', async () => {
      expect(response.status).toBe(204);
    });
    it('should update the user in the database', () => {
      expect(updatedUser[0].password).toBe(updateData.password);
    });
  });
  describe('GET /user:id', () => {
    let response: any;
    let dbUser: any;
    beforeAll(async () => {
      const query = { _id: userId };
      response = await request(server).get(`/user/${userId}?fields=email,password`);
      const [{ email, password }] = await User.find(query);
      dbUser = { email, password };
    });
    it('should return user & match snapshot', async () => {
      expect(stripIdAndNext(response.text)).toMatchSnapshot();
      expect(response.status).toBe(200);
    });
    it('shuld match with the user in the database', () => {
      expect(_.omit(JSON.parse(response.text).data, '_id')).toEqual(dbUser);
    });
  });
  describe('DELETE /user/:id', () => {
    let response: any;
    let dbResponse: any;
    beforeAll(async () => {
      const query = { _id: userId };
      response = await request(server).delete(`/user/${userId}`);
      dbResponse = await User.find(query);
    });
    it('should respond with 204', async () => {
      expect(response.status).toBe(204);
    });
    it('should remove the user from the database', () => {
      expect(dbResponse.length).toBe(0);
    });
  });
});
