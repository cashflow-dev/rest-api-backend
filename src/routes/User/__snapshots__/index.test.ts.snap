// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/user GET /user should get users 1`] = `"{\\"message\\":\\"ok\\",\\"data\\":[{\\"email\\":\\"test4@test.com\\",\\"password\\":\\"hashed\\"},{\\"email\\":\\"test3@test.com\\",\\"password\\":\\"hashed\\"},{\\"email\\":\\"test2@test.com\\",\\"password\\":\\"hashed\\"},{\\"email\\":\\"test1@test.com\\",\\"password\\":\\"hashed\\"},{\\"email\\":\\"test0@test.com\\",\\"password\\":\\"hashed\\"}],\\"next\\":\\"exists\\",\\"statusCode\\":200}"`;

exports[`/user GET /user:id should return user & match snapshot 1`] = `"{\\"message\\":\\"ok\\",\\"data\\":{\\"email\\":\\"test0@test.com\\",\\"password\\":\\"newTestPassword\\"},\\"statusCode\\":200}"`;

exports[`/user POST /user with all fields should match snapshot 1`] = `"Created"`;

exports[`/user POST /user with fields missing with email missing should match snapshot 1`] = `"{\\"statusCode\\":400,\\"errorCode\\":400,\\"errors\\":[\\"\\\\\\"email\\\\\\" is required\\"],\\"message\\":\\"ValidationError\\"}"`;

exports[`/user POST /user with fields missing with password missing should match snapshot 1`] = `"{\\"statusCode\\":400,\\"errorCode\\":400,\\"errors\\":[\\"\\\\\\"password\\\\\\" is required\\"],\\"message\\":\\"ValidationError\\"}"`;
